implicit/lmf.pyx

```diff
@@ -239,26 +246,26 @@ def lmf_update(RNGVector rng, floating[:, :] deriv_sum_sq,
                     i = indices[index]
                     for _ in range(n_factors):
                         exp_r += user_vectors[u, _] * item_vectors[i, _]
-                    exp_r = exp(exp_r)
-                    z = (data[index] * exp_r) / (1 + exp_r)
+                    z = sigmoid(exp_r) * data[index]
                     for _ in range(n_factors):
-                        deriv[_] -= z * item_vectors[i, _]
-
+                        deriv[_] =  deriv[_] - z * item_vectors[i, _]
+                
                 # Negative(Sampled) Item Indices exp(y_ui) / (1 + exp(y_ui)) * y_i
                 for _ in range(min(n_items, user_seen_item * neg_prop)):
                     index = rng.generate(thread_id)
                     i = indices[index]
                     exp_r = 0
                     for _ in range(n_factors):
-                        exp_r += user_vectors[u, _] * item_vectors[i, _]
-                    exp_r = exp(exp_r)
-                    z = exp_r / (1 + exp_r)
-                    for _ in range(n_factors):
-                        deriv[_] -= z * item_vectors[i, _]
+                        exp_r = exp_r + (user_vectors[u, _] * item_vectors[i, _])
+                    z = sigmoid(exp_r)
 
+                    for _ in range(n_factors):
+                        deriv[_] = deriv[_] - z * item_vectors[i, _]
                 for _ in range(n_factors):
                     deriv[_] -= reg * user_vectors[u, _]
                     deriv_sum_sq[u, _] += deriv[_] * deriv[_]
-                    user_vectors[u, _] += (lr / sqrt(deriv_sum_sq[u, _])) * deriv[_]
+
+                    # a small constant is added for numerical stability
+                    user_vectors[u, _] += (lr / (sqrt(1e-6 + deriv_sum_sq[u, _]))) * deriv[_]
```

Comment:
this line would be better

```python
    user_vectors[u, _] += (lr / (1 + sqrt(1e-6+ deriv_sum_sq[u, _]))) * deriv[_]
```

Reply:
not 1, but 1.0

Reply:
Why +1 ? I just added a small number (an "epsilon"), adding a higher constant such as 1 would bias it towards smaller learning rates or not?

Reply:
yes but original paper suggests tha(cant find original paper link) but. https://www.slideshare.net/mobile/ssuser77b8c6/an-overview-of-gradient-descent-optimization-algorithms


I think it makes sense because if we dont add 1, then learning rate at first few steps becomes extremely large because it divides sqrt of sum of square of grads(nearly 0 mostly)

Reply:
making learning rate larger than what we initialized

Reply:
I still can't follow you: IMHO 1 is *not* a small constant (especially, for tuning a learning rate). Even your link suggests 1e-8 (=0.00000001) as constant and does not mention the addition of 1 at all. I also looked at other sources (e.g., https://d2l.ai/chapter_optimization/adagrad.html) and there they also use a constant of 1e-6. I would understand it to use a smaller one than 1e-6, but 1.0 is one million times higher than other sources suggest.

Reply:
You are right, 1 is super larger number...

https://arxiv.org/pdf/1609.04747 suggests 1e-8

the default implementation in PyTorch suggests 1-e10
https://pytorch.org/docs/stable/optim.html?highlight=adagrad#torch.optim.Adagrad

btw, those implementations place epsilon out of square root and I'm not sure putting epsilon inside of sqrt is good or not(anyway would be much better before)